File: ambeth/Ambeth.Util/ambeth/collections/EmptyMap.cs
using De.Osthus.Ambeth.Util;
using System;
using System.Collections.Generic;

namespace De.Osthus.Ambeth.Collections
{
    public static class EmptyMap<K, V>
    {
        public static IMap<K, V> Empty()
        {
            // Note that the static type is only instantiated when
            // it is needed, and only then is the T[0] object created, once.
            return EmptyMapIntern.Instance;
        }

        private sealed class EmptyMapIntern : IMap<K, V>, IImmutableType
        {
            public static readonly EmptyMapIntern Instance = new EmptyMapIntern();
            
            public void Clear()
            {
                throw new NotSupportedException("Set is read-only");
            }

            public IISet<K> KeySet()
            {
                return EmptySet.Empty<K>();
            }

            public void KeySet(ICollection<K> targetKeySet)
            {
                // intended blank
            }

            public IList<K> KeyList()
	        {
		        return EmptyList.Empty<K>();
	        }

            public int Count
            {
                get
                {
                    return 0;
                }
            }

            public bool IsEmpty()
            {
                return true;
            }

            public IList<V> Values()
            {
                return EmptyList.Empty<V>();
            }

            public bool ContainsKey(K key)
            {
                return false;
            }

            public V Get(K key)
            {
                return default(V);
            }

            public K GetKey(K key)
            {
                return key;
            }

            public V Put(K key, V value)
            {
                throw new NotSupportedException("Set is read-only");
            }

            public bool PutIfNotExists(K key, V value)
            {
                throw new NotSupportedException("Set is read-only");
            }

            public V Remove(K key)
            {
                throw new NotSupportedException("Set is read-only");
            }

            public bool RemoveIfValue(K key, V value)
            {
                throw new NotSupportedException("Set is read-only");
            }

            public V[] ToArray()
            {
                return EmptyList.EmptyArray<V>();
            }

            public V[] ToArray(V[] targetArray)
            {
                return targetArray;
            }

            public Iterator<Entry<K, V>> Iterator()
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }

            public Iterator<Entry<K, V>> Iterator(bool removeAllowed)
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }

            public IEnumerator<Entry<K, V>> GetEnumerator()
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }

            System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }

            Iterator Iterable.Iterator()
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }

            Iterator Iterable.Iterator(bool removeAllowed)
            {
                return EmptyIterator<Entry<K, V>>.Empty<Entry<K, V>>();
            }
        }
    }
}

File: jambeth/jambeth-util/src/main/java/com/koch/ambeth/util/collections/EmptyMap.java
package com.koch.ambeth.util.collections;

import java.lang.reflect.Array;
import java.util.Collection;
import java.util.Iterator;
import java.util.Map;

import com.koch.ambeth.util.IImmutableType;

public final class EmptyMap<K, V> implements ILinkedMap<K, V>, IImmutableType
{
	public static final class EmptyMapIterator<K, V> implements Iterator<Entry<K, V>>
	{
		@SuppressWarnings("rawtypes")
		private static final EmptyMapIterator iteratorInstance = new EmptyMapIterator();

		@SuppressWarnings("unchecked")
		static <K, V> EmptyMapIterator<K, V> emptyIterator()
		{
			return iteratorInstance;
		}

		private EmptyMapIterator()
		{
			// intended blank
		}

		@Override
		public boolean hasNext()
		{
			return false;
		}

		@Override
		public Entry<K, V> next()
		{
			throw new UnsupportedOperationException();
		}

		@Override
		public void remove()
		{
			throw new UnsupportedOperationException();
		}

	}

	@SuppressWarnings("rawtypes")
	private static final EmptyMap instance = new EmptyMap();

	@SuppressWarnings("unchecked")
	public static <K, V> EmptyMap<K, V> emptyMap()
	{
		return instance;
	}

	private EmptyMap()
	{
		// intended blank
	}

	@Override
	public int size()
	{
		return 0;
	}

	@Override
	public boolean containsValue(Object value)
	{
		return false;
	}

	@Override
	public void putAll(Map<? extends K, ? extends V> m)
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@Override
	public Iterator<Entry<K, V>> iterator()
	{
		return EmptyMapIterator.emptyIterator();
	}

	@Override
	public Iterator<Entry<K, V>> iterator(boolean removeAllowed)
	{
		return EmptyMapIterator.emptyIterator();
	}

	@Override
	public void clear()
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@Override
	public boolean containsKey(Object key)
	{
		return false;
	}

	@Override
	public ISet<Entry<K, V>> entrySet()
	{
		return null;
	}

	@Override
	public void entrySet(ISet<Entry<K, V>> targetEntrySet)
	{
		// intended blank
	}

	@Override
	public V get(Object key)
	{
		return null;
	}

	@Override
	public boolean isEmpty()
	{
		return true;
	}

	@Override
	public ISet<K> keySet()
	{
		return EmptySet.emptySet();
	}

	@Override
	public void keySet(Collection<K> targetKeySet)
	{
		// intended blank
	}

	@Override
	public IList<K> keyList()
	{
		return EmptyList.<K> getInstance();
	}

	@Override
	public V put(K key, V value)
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@Override
	public V remove(Object key)
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@Override
	public IList<V> values()
	{
		return EmptyList.<V> getInstance();
	}

	@Override
	public K getKey(K key)
	{
		return null;
	}

	@Override
	public boolean putIfNotExists(K key, V value)
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@Override
	public boolean removeIfValue(K key, V value)
	{
		throw new UnsupportedOperationException("Map is read-only");
	}

	@SuppressWarnings("unchecked")
	@Override
	public V[] toArray(Class<V> arrayType)
	{
		return (V[]) Array.newInstance(arrayType, 0);
	}
}
